server/services/objectAssetService.js

```diff
@@ -62,28 +66,69 @@ async function getAssetsForArtworks(artworks) {
   // so we do need archival renditions as part of our list
   // and need some extra work to do so
-  const { objectNumber } = artworksInformation[0];
+  const artwork = { ...artworks[0] };
+  const artworkWithDAMSInformation = await addAssetFields(artwork);
+
+  return [artworkWithDAMSInformation];
+}
+
+async function getEnsembleImageUrl(ensembleIndex) {
+  const ensembleCacheKey = makeEnsembleCacheKey(ensembleIndex);
+  const ensembleImageUrl = memoryCache.get(ensembleCacheKey);
+
+  // If we don't have a cached version of this ensemble image url
+  // let's fetch it live, store it, then return it
+  if (!ensembleImageUrl) {
+    const liveEnsembleImageUrl = await damsService.getEnsembleImageUrl(
+      ensembleIndex
+    );
+    memoryCache.put(ensembleCacheKey, liveEnsembleImageUrl, oneWeek);
+
+    return liveEnsembleImageUrl;
+  }
+
+  // Otherwise, we have the cached version
+  return ensembleImageUrl;
+}
+
+/** Utility function to integrate new fields into the artwork object
+ * using information provided by rendition assets from the DAMS
+ *
+ * Fields added include
+ * - Renditions list from DAMS, including Archival Images
+ * - Ensemble Image URL for the ensemble image that lives in NetX
+ * - Published Provenance text
+ * - Published Archives Reference text
+ */
+async function addAssetFields(artwork) {
+  // Fetch the related asset from the DAMS
+  const objectNumber = artwork._source.invno ? artwork._source.invno : null;
   const artworkAssets = await getAssetByObjectNumber(objectNumber);
 
   // We store the renditions but also aditional fields needed
   // for single artwork rendering
-  return artworks.map((artwork) => {
-    const renditions = artworkAssets || [];
-    const rendition = renditions[0];
-
-    artwork._source["renditions"] = renditions;
-    artwork._source["publishedProvenance"] = rendition
-      ? damsService.getValueFromAsset("Published Provenance (TMS)", rendition)
-      : "";
-    artwork._source["publishedArchivesReference"] = rendition
-      ? damsService.getValueFromAsset(
-          "Published Archives Reference (TMS)",
-          rendition
-        )
-      : "";
-
-    return artwork;
-  });
+  const renditions = artworkAssets || [];
+  const rendition = renditions[0];
+
+  // Get the ensemble image url to use from the DAMS
+  const ensembleImageUrl = artwork._source.ensembleIndex
+    ? await getEnsembleImageUrl(artwork._source.ensembleIndex)
+    : null;
+
+  // Add new fields based on information from the DAMS
+  artwork._source["ensembleImageUrl"] = ensembleImageUrl;
+  artwork._source["renditions"] = renditions;
+  artwork._source["publishedProvenance"] = rendition
+    ? damsService.getValueFromAsset("Published Provenance (TMS)", rendition)
+    : "";
+  artwork._source["publishedArchivesReference"] = rendition
+    ? damsService.getValueFromAsset(
+        "Published Archives Reference (TMS)",
+        rendition
+      )
+    : "";
+
+  return artwork;
```

Comment:
Mainly extracted the logic from the left-side into this new `addAssetFields` function for better code encapsulation